### PR TITLE
Rewrite the core of the binding generator.

### PR DESCRIPTION
**Pull request by emilio**
TL;DR: The binding generator is a mess as of right now. At first it was funny
(in a "this is challenging" sense) to improve on it, but this is not
sustainable.

The truth is that the current architecture of the binding generator is a huge
pile of hacks, so these few days I've been working on rewriting it with a few
goals.

 1) Have the hacks as contained and identified as possible. They're sometimes
    needed because how clang exposes the AST, but ideally those hacks are well
    identified and don't interact randomly with each others.

    As an example, in the current bindgen when scanning the parameters of a
    function that references a struct clones all the struct information, then if
    the struct name changes (because we mangle it), everything breaks.

 2) Support extending the bindgen output without having to deal with clang. The
    way I'm aiming to do this is separating completely the parsing stage from
    the code generation one, and providing a single id for each item the binding
    generator provides.

 3) No more random mutation of the internal representation from anywhere. That
    means no more `Rc<RefCell<T>>`, no more random circular references, no more
    borrow_state... nothing.

 4) No more deduplication of declarations before code generation.

    Current bindgen has a stage, called `tag_dup_decl`[1], that takes care of
    deduplicating declarations. That's completely buggy, and for C++ it's a
    complete mess, since we YOLO modify the world.

    I've managed to take rid of this using the clang canonical declaration, and
    the definition, to avoid scanning any type/item twice.

 5) Code generation should not modify any internal data structure. It can lookup
    things, traverse whatever it needs, but not modifying randomly.

 6) Each item should have a canonical name, and a single source of mangling
    logic, and that should be computed from the immutable state, at code
    generation.

    I've put a few canonical_name stuff in the code generation phase, but it's
    still not complete, and should change if I implement namespaces.

Improvements pending until this can land:

 1) Add support for missing core stuff, mainly generating functions (note that
    we parse the signatures for types correctly though), bitfields, generating
    C++ methods.

 2) Add support for the necessary features that were added to work around some
    C++ pitfalls, like opaque types, etc...

 3) Add support for the sugar that Manish added recently.

 4) Optionally (and I guess this can land without it, because basically nobody
    uses it since it's so buggy), bring back namespace support.

These are not completely trivial, but I think I can do them quite easily with
the current architecture.

I'm putting the current state of affairs here as a request for comments...
 Note that there are still a few smells I want to eventually
re-redesign, like the ParseError::Recurse thing, but until that happens I'm
way happier with this kind of architecture.

I'm keeping the old `parser.rs` and `gen.rs` in tree just for reference while I
code, but they will go away.

[1]: https://github.com/Yamakaky/rust-bindgen/blob/master/src/gen.rs#L448

### CODE REVIEW

**Comment by emilio** [comment]
cc: @nox, @bholley, @Manishearth, @vvuk, @Yamakaky, any thoughts on this approach?

File: src/gen.rs [hunk 1]
```diff
@@ -989,9 +989,9 @@ fn cstruct_to_rs(ctx: &mut GenCtx, name: &str, ci: CompInfo,
             });
         }
 
-        // FIXME: rustc actually generates tons of warnings
-        // due to an empty repr(C) type, so we just generate
-        // a dummy field with pointer-alignment to supress it.
+        // FIXME: rustc actually generates tons of warnings due to an empty
+        // repr(C) type, so we just generate a dummy field with pointer
+        // alignment to supress it.
```
Comment (nox) [inline src/gen.rs]: What about `Foo(c_void)`?

**Comment by nox** [comment]
This looks great!

File: src/gen.rs (hunk 1 above)
Comment (emilio) [inline src/gen.rs]: Wfm, this is just an accidental comment rewrapping fwiw.

**Comment by Yamakaky** [comment]
Yes! The `Rc<Borrow>` especially looks awful.
While we are at it, it would be cool to check if all the features in https://github.com/Yamakaky/rust-bindgen are present in this fork. That way, we could close the other and use this one. (see https://github.com/servo/rust-bindgen/issues/21)

**Comment by Yamakaky** [comment]
Also, I made some refactoring myself, it could help you

**Comment by emilio** [comment]
@Yamakaky I planned to focus in the main architecture change.

Using clang-rs vs our hand-rolled thing would be great (never heard of that library before, but looks nice). That being said, this patch is already ~4k lines, and I don't want to make it larger unnecessarily.

Other parts that I don't plan to touch right now is argument parsing and such, feel free to help with it if you feel like it :-)

**Comment by Yamakaky** [comment]
Maybe clang-rs should be integrated before you make too many changes. It would have been work for nothing.

**Comment by emilio** [comment]
Clang-rs is only a wrapper for types like `Cursor` or `Type`, it doesn't significantly change the architecture.

**Comment by Yamakaky** [comment]
(inviting @KyleMayes in the conversation)

**Comment by emilio** [comment]
Current status: all the tests produce compilable code, I've been able to generate bindings for gecko without crashing, parsing all the relevant stl stuff and mfbt (obviously, the resulting code didn't compile due to sfinae and similar stuff).

I hope to be able to fully support c++ namespaces, though I'm going to keep working in getting the features up to date with current bindgen before doing more tests like that.

**Comment by emilio** [comment]
Hey, I've heard over there that this passes some tests already :-)

**Comment by emilio** [comment]
This is basically ready for review.

r? @nox

**Comment by emilio** [comment]
Stylo PR with example script and generated code for Gecko: https://github.com/servo/servo/pull/13269

File: Cargo.toml [hunk 2]
```diff
@@ -1,8 +1,12 @@
...
+  "Emilio Cobos Álvarez <ecoal95@gmail.com>",
+  "The Servo project developers",
+]
 build = "build.rs"
 description = "A binding generator for Rust"
-homepage = "https://github.com/crabtw/rust-bindgen"
+homepage = "https://github.com/servo/rust-bindgen"
 keywords = ["bindings", "ffi", "code-generation"]
 license = "BSD-3-Clause"
 name = "bindgen"
```
Comment (Manishearth) [inline Cargo.toml]: Probably should bump to 0.17.0 or even 1.0.0

**Comment by nox** [comment]
Why do we bump the version even though we didn't rename the crate? Either we should follow upstream's version or we should rename this fork to servo-bindgen.

File: methods.hpp [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class Foo {
+public:
+  void bar();
+  void bar(int);
+  static void baz();
+};
```
Comment (nox) [inline methods.hpp]: What is the purpose of this file?

File: src/clang.rs [hunk 4]
```diff
@@ -15,9 +15,24 @@ pub struct Cursor {
     x: CXCursor
 }
 
+impl fmt::Debug for Cursor {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "Cursor({} kind: {}, loc: {})",
+               self.spelling(), kind_to_str(self.kind()), self.location())
```
Comment (nox) [inline src/clang.rs]: Nit: maybe use `Formatter::debug_tuple`.

File: src/codegen/mod.rs [hunk 5]
```diff
@@ -0,0 +1,2000 @@
...
+    let root = ctx.root_module().canonical_name(ctx);
+    let root_ident = ctx.rust_ident(&root);
+    quote_item!(ctx.ext_cx(), use $root_ident;).unwrap()
+}
+
+macro_rules! repr {
+    ($which:expr) => {{
+        aster::AstBuilder::new().attr().list("repr").words(&[$which]).build()
+    }}
+}
```
Comment (nox) [inline src/codegen/mod.rs]: This could be a plain old function.

File: src/codegen/mod.rs [hunk 6]
```diff
@@ -0,0 +1,2000 @@
...
+
+macro_rules! doc {
+    ($comment:expr) => {{
+        aster::AstBuilder::new().attr().doc($comment)
+    }}
+}
```
Comment (nox) [inline src/codegen/mod.rs]: This could be a plain old function.

File: src/codegen/mod.rs [hunk 7]
```diff
@@ -0,0 +1,2000 @@
...
+
+macro_rules! link_name {
+    ($name:expr) => {{
+        aster::AstBuilder::new().attr().name_value("link_name").str($name)
+    }}
+}
```
Comment (nox) [inline src/codegen/mod.rs]: This could be a plain old function.

File: src/codegen/mod.rs [hunk 8]
```diff
@@ -0,0 +1,2000 @@
...
+
+    fn seen(&self, item: ItemId) -> bool {
+        self.items_seen.contains(&item)
+    }
+
+    fn set_seen(&mut self, item: ItemId) {
+        self.items_seen.insert(item);
+    }
+
+    fn seen_function(&self, name: &String) -> bool {
```
Comment (nox) [inline src/codegen/mod.rs]: Why `&String`?

File: src/codegen/mod.rs [hunk 9]
```diff
@@ -0,0 +1,2000 @@
...
+    }
+}
+
+impl ops::DerefMut for CodegenResult {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.items
+    }
+}
+
+struct ForeignModBuilder {
```
Comment (nox) [inline src/codegen/mod.rs]: Could all of the builders not specific to bindgen live in their own module?

File: src/codegen/mod.rs [hunk 10]
```diff
@@ -0,0 +1,2000 @@
...
+            span: DUMMY_SP,
+        })
+    }
+}
+
+struct ArrayTyBuilder {
+    len: usize,
+}
+
+// TODO: Aster seem to lack a array builder?
```
Comment (nox) [inline src/codegen/mod.rs]: File an issue about this.

File: src/codegen/mod.rs [hunk 11]
```diff
@@ -0,0 +1,2000 @@
...
+
+        P(ast::Ty {
+            id: ast::DUMMY_NODE_ID,
+            node: array_kind,
+            span: DUMMY_SP,
+        })
+    }
+}
+
+struct BlobTyBuilder {
```
Comment (nox) [inline src/codegen/mod.rs]: It is not obvious what this one does.

File: src/codegen/mod.rs [hunk 12]
```diff
@@ -0,0 +1,2000 @@
...
+    }
+}
+
+/// A trait to convert a rust type into a pointer, optionally const, to the same
+/// type.
+///
+/// This is done due to aster's lack of pointer builder, I guess I should PR
+/// there.
+trait ToPtr {
+    fn to_ptr(self, is_const: bool, span: Span) -> P<ast::Ty>;
```
Comment (nox) [inline src/codegen/mod.rs]: Use `ast::Mutability` for the `is_const` argument.

File: src/codegen/mod.rs [hunk 13]
```diff
@@ -0,0 +1,2000 @@
...
+        });
+        P(ast::Ty {
+            id: ast::DUMMY_NODE_ID,
+            node: ty,
+            span: span,
+        })
+    }
+}
+
+trait CodeGenerator {
```
Comment (nox) [inline src/codegen/mod.rs]: Bikeshed: this should be called `Codegen`.

File: src/codegen/mod.rs [hunk 14]
```diff
@@ -0,0 +1,2000 @@
...
+
+    fn codegen(&self,
+               ctx: &BindgenContext,
+               result: &mut CodegenResult,
+               item: &Item) {
+        use aster::struct_field::StructFieldBuilder;
+        // Don't output classes with template parameters that aren't types, and
+        // also don't output template specializations, neither total or partial.
+        //
+        // TODO: Generate layout tests for template specializations, yay!
```
Comment (nox) [inline src/codegen/mod.rs]: File an issue about that.

File: src/codegen/mod.rs [hunk 15]
```diff
@@ -0,0 +1,2000 @@
...
+
+        if ty.can_derive_copy(ctx) {
+            derives.push("Copy");
+            if !applicable_template_args.is_empty() {
+                // FIXME: This requires extra logic if you have a big array in a
+                // templated struct. The reason for this is that the magic:
+                //     fn clone(&self) -> Self { *self }
+                // doesn't work for templates.
+                //
+                // It's not hard to fix though.
```
Comment (nox) [inline src/codegen/mod.rs]: File an easy issue? :)

File: src/codegen/mod.rs [hunk 16]
```diff
@@ -0,0 +1,2000 @@
...
+
+        let sig = ast::MethodSig {
+            unsafety: ast::Unsafety::Unsafe,
+            abi: Abi::Rust,
+            decl: P(fndecl.clone()),
+            generics: ast::Generics::default(),
+            constness: ast::Constness::NotConst,
+        };
+
+        // TODO: This needs to be kept in sync, so we'd be better unifying this,
```
Comment (nox) [inline src/codegen/mod.rs]: Kept in sync with what?

File: src/codegen/mod.rs [hunk 17]
```diff
@@ -0,0 +1,2000 @@
...
+
+type ItemSet = BTreeSet<ItemId>;
+
+trait TypeCollector {
+    type Extra;
+
+    fn collect_types(&self,
+                     context: &BindgenContext,
+                     types: &mut ItemSet,
+                     extra: &Self::Extra);
```
Comment (nox) [inline src/codegen/mod.rs]: Wouldn't it be simpler to pass `Self::Extra` by value?

File: src/codegen/mod.rs [hunk 18]
```diff
@@ -0,0 +1,2000 @@
...
+                        if context.options().whitelisted_vars.matches(&name) {
+                            items.insert(item.id());
+                            var.ty().collect_types(context, &mut items, &());
+                        }
+                    }
+                    ItemKind::Module(..) => {}
+                }
+            }
+
+            fn contains_parent(ctx: &BindgenContext, types: &ItemSet, id: ItemId) -> bool {
```
Comment (nox) [inline src/codegen/mod.rs]: Bikeshed: `is_recursive`?

File: src/ir/comp.rs [hunk 19]
```diff
@@ -0,0 +1,718 @@
...
+
+    pub fn comment(&self) -> Option<&str> {
+        self.comment.as_ref().map(|c| &**c)
+    }
+
+    pub fn bitfield(&self) -> Option<u32> {
+        self.bitfield
+    }
+
+    pub fn mutable(&self) -> bool {
```
Comment (nox) [inline src/ir/comp.rs]: Bikeshed: `is_mutable`?

File: src/ir/annotations.rs [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+        anno.parse(&cursor.comment(), &mut matched_one);
+
+        if matched_one {
+            Some(anno)
+        } else {
+            None
+        }
+    }
+
+    pub fn hide(&self) -> bool {
```
Comment (nox) [inline src/ir/annotations.rs]: Bikeshed: `is_hidden`?

File: src/ir/annotations.rs [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+        self.hide
+    }
+
+    pub fn opaque(&self) -> bool {
```
Comment (nox) [inline src/ir/annotations.rs]: Bikeshed: `is_opaque`?

File: src/ir/annotations.rs [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+        self.opaque
+    }
+
+    pub fn use_as(&self) -> Option<&str> {
```
Comment (nox) [inline src/ir/annotations.rs]: I don't understand this method.

File: src/ir/annotations.rs [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+        self.use_as.as_ref().map(|s| &**s)
+    }
+
+    pub fn no_copy(&self) -> bool {
```
Comment (nox) [inline src/ir/annotations.rs]: I don't understand this method, and the name should probably not be a negation.

File: src/ir/annotations.rs [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+        self.no_copy
+    }
+
+    pub fn private_fields(&self) -> Option<bool> {
```
Comment (nox) [inline src/ir/annotations.rs]: Bikeshed: `has_private_fields`?

File: methods.hpp (hunk 3 above)
Comment (emilio) [inline methods.hpp]: Leftover file, good catch. Removing...

File: src/codegen/mod.rs (hunk 8 above)
Comment (emilio) [inline src/codegen/mod.rs]: Bah, I wrote this on the fly and I assumed that the `contains` method would be harcoded to take `&Key`. Thanks for the catch :)

File: src/codegen/mod.rs (hunk 9 above)
Comment (emilio) [inline src/codegen/mod.rs]: Sure, I'll move most of this to a `codegen/builders`, `codegen/helpers`, or similar.

File: src/codegen/mod.rs (hunk 10 above)
Comment (emilio) [inline src/codegen/mod.rs]: Filled https://github.com/serde-rs/aster/issues/104

File: src/codegen/mod.rs (hunk 11 above)
Comment (emilio) [inline src/codegen/mod.rs]: It creates an appropriate type for a blob with a given `Layout`, with proper size and alignment. But I'll rather write this in a comment.

File: src/codegen/mod.rs (hunk 12 above)
Comment (emilio) [inline src/codegen/mod.rs]: Bindgen only tracks whether a field is const or not, so while I agree it would be a cleaner interface, I think changing all the callsites from:

```rust
ty.to_ptr(item.is_const(), ctx.span())
```

to 

```rust
ty.to_ptr(if item.is_const() { ast::Mutability::Mutable } else { ast::Mutability::Immutable }, ctx.span())
```

is counter-productive. I could add a method to the item that returns the actual mutability, but that means either introducing the `ir` module to `syntax::ast` code, or creating another artificial trait to implement it there.

File: src/codegen/mod.rs (hunk 16 above)
Comment (emilio) [inline src/codegen/mod.rs]: With the other loop that decides argument naming. Clarified it in the comment.

File: src/codegen/mod.rs (hunk 17 above)
Comment (emilio) [inline src/codegen/mod.rs]: At first I thought so, but that implies either parameterizing the trait and type on a lifetime and whatnot, or pass `Item`s by value, which is both inefficient and not the intended thing. I tried to keep this simple since `extra` uses to be just the `Item` that refers to us, but if that changes we can refactor the interface.

File: src/ir/comp.rs (hunk 19 above)
Comment (emilio) [inline src/ir/comp.rs]: sure :)

File: src/ir/annotations.rs (hunk 22 above)
Comment (emilio) [inline src/ir/annotations.rs]: This is used to implement the `replaces` annotation. I've given it a more meaningful name and a doc comment :)

File: src/ir/annotations.rs (hunk 24 above)
Comment (emilio) [inline src/ir/annotations.rs]: This does not indicate whether a type *has* private fields, but whether it *should generate* private fields. I'll give it a meaningful name.

**Comment by emilio** [comment]
I addressed a bunch of review comments, and cleaned up a bit more. Pushed as to-squash commits to ease the review.

File: src/codegen/mod.rs (hunk 12 above)
Comment (nox) [inline src/codegen/mod.rs]: OK.

File: src/codegen/mod.rs (hunk 17 above)
Comment (nox) [inline src/codegen/mod.rs]: OK.

File: src/codegen/mod.rs (hunk 18 above)
Comment (nox) [inline src/codegen/mod.rs]: OK.

**Comment by emilio** [comment]
review ping @nox, is this good to land?
